JDK-8301749: Tracking malloc pooled memory size

jdksjolen opened the pull request:
This PR implements `os::malloc_info` by calling into glibc's `malloc_info` and provides a JCmd for asking the JVM to call `malloc_info` and print out the results. The user can make an informed decision regarding heap trimming by parsing the output of the JCmd. I wanted to avoid parsing the XML inside of the JVM , as this functionality may turn out to be "good enough". A future enhancement would be for the JVM to make this informed decision by parsing the output. Another may be to present this with other data, for example in NMT.

@tstuefe, I think you'd be interested in this PR as it builds upon your `trim_native_heap` functionality.

I'm using weak symbols (as per ELF) in order to compile this for both GLibc and Musl.

The original bug description is this:

>Memory not allocated via NMT and uncommitted memory contributes to a discrepancy between NMT and RSS.
It can be hard to distinguish between these cases.
If the VM can determine the amount of pooled memory by malloc it would help with both determining if a trim_native_memory is needed, or there are some allocations happening outside of NMT.
>
>It would be good if the VM can summarize the malloc pooled memory by using e.g. malloc_info(3).
>
>We can thus more precise account for RSS value.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8302128](https://bugs.openjdk.org/browse/JDK-8302128) to be approved

### Issues
 * [JDK-8301749](https://bugs.openjdk.org/browse/JDK-8301749): Tracking malloc pooled memory size
 * [JDK-8302128](https://bugs.openjdk.org/browse/JDK-8302128): Tracking malloc pooled memory size (**CSR**)


### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [5d93207c](https://git.openjdk.org/jdk/pull/12455/files/5d93207c34a5e355c07796ae7217d2225bdd5098)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12455/head:pull/12455` \
`$ git checkout pull/12455`

Update a local copy of the PR: \
`$ git checkout pull/12455` \
`$ git pull https://git.openjdk.org/jdk pull/12455/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12455`

View PR using the GUI difftool: \
`$ git pr show -t 12455`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12455.diff">https://git.openjdk.org/jdk/pull/12455.diff</a>

</details>

jdksjolen commented:
I'm opening this PR as a suggested change. I'm not strongly attached to the idea that it must be a JCmd, or that we must simply return the info as it is. I am only strongly attached to the idea that the user should have an easy way of receiving the information that `malloc_info` offers. If anyone has any suggestions, please share them!

poonamparhar commented:
@jdksjolen would be good to share the output of this command. Would give an idea to the reviewers as to what information gets printed with this command.

robehn commented:
I'm wondering if we should try to warn in the case when user have preloaded a thrid party malloc that don't have malloc_info. But when we load libc that symbol will find a target.
If it's not to much trouble it would be pretty nice if we could return some output to the user indicating this.
Since there a few ways to high-jack malloc maybe it's hard to figure out and argubly if the user preloaded something he should be aware of it, but maybe not aware if it contains malloc_info or not.

dholmes-ora commented on src/hotspot/os/linux/mallocInfoDcmd.cpp:
```diff
@@ -0,0 +1,48 @@
+#include <malloc.h>
+
+void MallocInfoDcmd::execute(DCmdSource source, TRAPS) {
+#ifdef LINUX
```
This is a linux-only file. ???

tstuefe replied:
Yes, this is all Linux. Instead, guard with GLIBC, and print in the else branch something like "not a glibc platform".

dholmes-ora commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -217,6 +220,13 @@ julong os::available_memory() {
 }
 
+int os::Linux::malloc_info(FILE* stream) {
+  if (::malloc_info) {
```
This looks strange as an existence check for a library function. At build time what does this get compiled into such that it runtime it can determine whether `malloc_info` exists or not ??

(also style rules say no implicit booleans so check against nullptr).

tstuefe replied:
I agree, this looks weird. Please do as the Romans do (in this case, as we do with mallinfo/mallinfo2()). Just follow the breadcrumbs.

Additionally, I would put this alongside os::Linux::get_mallinfo and guard it with ifdef GLIBC. You don't need this on muslc.

dholmes-ora reviewed (changes requested):
General idea is fine - thanks for implementing.

There is unnecessary ifdef usage, and also a query about the existence check, below.

This needs a jcmd documentation update - closed MR to update ` closed/src/jdk.jcmd/share/man/jcmd.md` and then include the updated jcmd.1 manpage file here.

tstuefe commented on src/hotspot/os/linux/os_linux.hpp:
```diff
@@ -193,6 +193,10 @@ class os::Linux {
+  // Calls out to GNU extension malloc_info if available
+  // otherwise does nothing and returns 0.
+  static int malloc_info(FILE* stream);
+
```
Could you put this at the end of the file, alongside os::get_mallinfo(), and within its ifdef GLIBC boundaries?

tstuefe commented on src/hotspot/os/linux/mallocInfoDcmd.cpp:
```diff
@@ -0,0 +1,48 @@
+#ifdef LINUX
+  char* buf;
+  size_t size;
+  ALLOW_C_FUNCTION(::open_memstream, FILE* stream = ::open_memstream(&buf, &size);)
```
I wondered how open_memstream communicates buffer growth (which may come with re-location). Then I realized that it just modifies the buffer start address the caller gives it in-place. What a weird API :)

You should handle error returns from open_memstream.

tstuefe commented on src/hotspot/os/linux/mallocInfoDcmd.cpp:
```diff
@@ -0,0 +1,48 @@
+  char* buf;
+  size_t size;
+  ALLOW_C_FUNCTION(::open_memstream, FILE* stream = ::open_memstream(&buf, &size);)
+  if (!os::Linux::malloc_info(stream)) {
```
Style nit, can you change this to `if (os::Linux::malloc_info(stream) == 0)` ? Using `!` it reads like the failure branch.

dholmes-ora replied:
Yes - again no implicit booleans

tstuefe commented on src/hotspot/os/linux/mallocInfoDcmd.cpp:
```diff
@@ -0,0 +1,48 @@
+ */
+
+#include "precompiled.hpp"
+#include "runtime/os.inline.hpp"
```
why this? why not os_linux.hpp?

tstuefe commented on src/hotspot/os/linux/mallocInfoDcmd.cpp:
```diff
@@ -0,0 +1,48 @@
+#include "precompiled.hpp"
+#include "runtime/os.inline.hpp"
+#include "mallocInfoDcmd.hpp"
+#include "utilities/debug.hpp"
```
What do you use from debug.hpp?

tstuefe commented on src/hotspot/os/linux/mallocInfoDcmd.hpp:
```diff
@@ -0,0 +1,51 @@
+public:
+  MallocInfoDcmd(outputStream* output, bool heap) : DCmd(output, heap) {}
+  static const char* name() {
+    return "System.malloc_info";
```
Nit, since the other one is called trim_native_heap, why not "native_heap_info"?

tstuefe commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -114,6 +114,9 @@
 #endif
 
+// Forward (re-)declare malloc_info in order to support Alpine Linux.
+int __attribute__((weak)) malloc_info(int options, FILE *stream);
```
Not needed if you do it like os::get_mallinfo() and dlsym the API as proposed.

tstuefe commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -217,6 +220,13 @@ julong os::available_memory() {
+  if (::malloc_info) {
+    return ::malloc_info(0, stream);
+  }
+  return 0;
```
return -1.

You aim to mimic the malloc_info() API at os:: level, so you should return -1 in case malloc_info failed, including failure do to resolving issues.

jdksjolen replied:
> return -1.
> 
> You aim to mimic the malloc_info() API at os:: level, so you should return -1 in case malloc_info failed, including failure do to resolving issues.

Wouldn't that lead to a caller not knowing whether `malloc_info` itself failed or if our wrapper failed? `malloc_info` also sets `errno`. I think it's better if we don't imitate `malloc_info`, by returning `-2` if `malloc_info` is missing, to distinguish the different states.

jdksjolen replied:
> > > return -1.
> > > You aim to mimic the malloc_info() API at os:: level, so you should return -1 in case malloc_info failed, including failure do to resolving issues.
> > 
> > 
> > Wouldn't that lead to a caller not knowing whether `malloc_info` itself failed or if our wrapper failed?
> 
> Why would that matter?
> 
This matters when a caller sees -1, reads errno, and then takes an action depending on errno. If it was our wrapper that failed, then this will be a "stale" errno and a faulty action will be taken. This matters in `MallocInfoDcmd::execute` because we send different error messages depending on whether the wrapper failed or `malloc_info(3)` failed.

tstuefe replied:
Okay, fair enough. I would have thought the user cannot really do anything about it anyway, but your way is certainly more precise.

tstuefe reviewed (changes requested):
@jdksjolen,

interesting. I considered malloc_info() too, but did not feel like parsing XML.

Generally, I would prefer if you keep it closer to what os::get_mallinf() is doing:

- only define and call that API for ifdef GLIBC
- resolve the traditional way via function pointers (see our mallinf()/mallinf2() wrappers)
- have the dcmd execute functions guared with GLIBC, and an else branch saying its not glibc

This also needs tests. See existing test for trim_native_heap. I would also like a test for muslc to test the stub functionality (its missing from TrimLibcHeapTest.java, which should be added).

jdksjolen commented:
Hi, here's a new version.

Notable changes:

1. Moved stuff around, used `dlsym()` and `__GLIBC__` as requested
2. More error handling
3. Rename the Jcmd to `System.native_heap_info`
4. Added a test

I left source code names with using `malloc_info` and not `native_heap_info`.We reference `malloc_info(3)` in the error messages and in the description and plainly deliver `malloc_info(3)`'s XML (and therefore a user must find their schema). Clearly this is not an abstraction over `malloc_info(3)`, so I thought it made sense to keep that clear in the source also.

@dholmes-ora, I'll add a closed PR when everything else has received thumbs up here.

dholmes-ora commented:
> @dholmes-ora, I'll add a closed PR when everything else has received thumbs up here.

Okay but note you need to integrate the jcmd.1 changes as part of this PR too.

Also (as @vidmik reminded me):

/csr needed

tstuefe commented on src/hotspot/os/linux/mallocInfoDcmd.cpp:
```diff
@@ -0,0 +1,61 @@
+  int err = os::Linux::malloc_info(stream);
+  if (err == 0) {
+    ALLOW_C_FUNCTION(::fflush, fflush(stream);)
+      _output->print_raw(buf);
```
indentation

tstuefe commented on src/hotspot/os/linux/mallocInfoDcmd.cpp:
```diff
@@ -0,0 +1,61 @@
+  ALLOW_C_FUNCTION(::fclose, ::fclose(stream);)
+  ALLOW_C_FUNCTION(::free, ::free(buf);)
+#else
+    _output->print(malloc_info_unavailable);
```
intendation

tstuefe commented on test/hotspot/jtreg/serviceability/dcmd/vm/MallocInfoTest.java:
```diff
@@ -0,0 +1,52 @@
+        OutputAnalyzer output = executor.execute("System.native_heap_info");
+        output.reportDiagnosticSummary();
+        output.shouldNotMatch(".*Error.*");
+        output.shouldMatch(".*<malloc version=.*");
```
small nit, you could probably get by with shouldContain here to save some cycles.

tstuefe reviewed (commented):
Small nits remain, otherwise fine.

dholmes-ora commented on test/hotspot/jtreg/serviceability/dcmd/vm/MallocInfoTest.java:
```diff
@@ -0,0 +1,52 @@
+public class MallocInfoTest {
+    public void run(CommandExecutor executor) {
+        OutputAnalyzer output = executor.execute("System.native_heap_info");
+        output.reportDiagnosticSummary();
```
Do this after the other checks, else on error you will get the report twice.

tstuefe replied:
I disagree, reportDiagnosticSummary is not useful after the tests: you only need it if the matches fail, to see the output being matched.

(I don't disagree strongly enough to block the patch though, if you want to go this way its fine too)

dholmes-ora replied:
Just to clarify something in relation to a  comment from @tstuefe that I now can't find. It can be useful to show the information even if the test passes, so that, for example, we can go and look at test results and see the kind of data that malloc_info is showing us. I often find myself adding this to tests because they swallow the interesting output even though the test passes (sometimes for the wrong reason!).

dholmes-ora commented on test/hotspot/jtreg/serviceability/dcmd/vm/MallocInfoTest.java:
```diff
@@ -0,0 +1,52 @@
+    public void run(CommandExecutor executor) {
+        OutputAnalyzer output = executor.execute("System.native_heap_info");
+        output.reportDiagnosticSummary();
+        output.shouldNotMatch(".*Error.*");
```
Can you test exitValue instead of parsing for an error?

jdksjolen replied:
I don't think so. Looking at  the `DCmd` class, there seems to be no support of specifying exit value.

tstuefe commented on src/hotspot/os/linux/mallocInfoDcmd.cpp:
```diff
@@ -30,24 +30,24 @@
   ALLOW_C_FUNCTION(::open_memstream, FILE* stream = ::open_memstream(&buf, &size);)
   if (stream == nullptr) {
-    _output->print("Error: Could not call malloc_info(3)");
+    _output->print("Error: Could not call malloc_info(3)\n");
```
Why not print_cr() ?

tstuefe commented on src/hotspot/os/linux/mallocInfoDcmd.cpp:
```diff
@@ -30,24 +30,24 @@
+    _output->print_raw(buf);
   } else if (err == -1) {
-    _output->print("Error: %s", os::strerror(errno));
+    _output->print("Error: %s\n", os::strerror(errno));
```
print_cr() ?

tstuefe commented on src/hotspot/os/linux/mallocInfoDcmd.cpp:
```diff
@@ -30,24 +30,24 @@
   if (err == 0) {
     ALLOW_C_FUNCTION(::fflush, fflush(stream);)
-      _output->print_raw(buf);
+    _output->print_raw(buf);
```
Append newline (`_output->cr()`)

dholmes-ora commented:
> Oh, certainly. But its even more useful the other way around, if the test fails. 

@tstuefe if a check fails then reportDiagnosticSummary is already called. E.g.
```
   public OutputAnalyzer shouldContain(String expectedString) {
        String stdout = getStdout();
        String stderr = getStderr();
        if (!stdout.contains(expectedString) && !stderr.contains(expectedString)) {
            reportDiagnosticSummary();
            throw new RuntimeException("'" + expectedString + "' missing from stdout/stderr");
        }
```

jdksjolen commented:
Hi,

I've updated the man page and done the `\n` -> `cr` conversion also. Please have a look :).

tstuefe commented on src/hotspot/os/linux/mallocInfoDcmd.cpp:
```diff
@@ -0,0 +1,62 @@
+  ALLOW_C_FUNCTION(::fclose, ::fclose(stream);)
+  ALLOW_C_FUNCTION(::free, ::free(buf);)
+#else
+  _output->print(malloc_info_unavailable);
```
Misses a cr

tstuefe reviewed (commented):
Looks good, apart from a missing newline in the musl path.

Did you test with musl?

dholmes-ora commented on src/jdk.jcmd/share/man/jcmd.1:
```diff
@@ -35,7 +35,7 @@
 . ftr VB CB
 . ftr VBI CBI
 .\}
-.TH "JCMD" "1" "2023" "JDK 21-ea" "JDK Commands"
```
Sorry my bad I gave you the wrong command to use to generate this. I should have been VERSION_SHORT=21-ea

jdksjolen replied:
No worries, I'll just revert the change.

jdksjolen commented:
> Looks good, apart from a missing newline in the musl path.
> 
> Did you test with musl?

I have not tested with musl. I'm attempting to cross-build but no success so far :). If you've got a working musl build then I'd be very thankful if you could test these changes.

dholmes-ora reviewed (approved):
Nothing further from me. Looks good. Thanks.

tstuefe commented:
> > Looks good, apart from a missing newline in the musl path.
> > Did you test with musl?
> 
> I have not tested with musl. I'm attempting to cross-build but no success so far :). If you've got a working musl build then I'd be very thankful if you could test these changes.

I'll give it a spin on my alpine box.

tstuefe commented:
Hi @jdksjolen, I tested musl and it works.

Please see suggested change to add regression testing for musl:

https://github.com/tstuefe/jdk/commit/130d68b5bc7b3dcccf27a365b916b9d158ca1f68

Cheers, Thomas

jdksjolen commented:
> Hi @jdksjolen, I tested musl and it works.
> 
> Please see suggested change to add regression testing for musl:
> 
> [tstuefe@130d68b](https://github.com/tstuefe/jdk/commit/130d68b5bc7b3dcccf27a365b916b9d158ca1f68)
> 
> Cheers, Thomas

Suggested changes looks good, I applied them. Thanks for testing this out.

Cheers

tstuefe reviewed (approved):
LGTM

jdksjolen commented:
Passes tier1.

/integrate
